draft-normington-jsonpath-latest.xml

```diff
@@ -109,37 +109,81 @@
+               a possibly empty list of nodes of the JSON document which match the JSONPath or an error (but not both).
             </t>
-            <t>If the JSONPath is empty, this corresponds to an empty sequence of matchers and matches the root node of the JSON document.
-               That is, the result is a list consisting of just the root node of the JSON document.
+            <t>If there is no matching node and no error has occurred, an implementation MUST return an empty list of nodes.
             </t>
-            <t>As a consequence, if any of the matchers matches no nodes, then the whole JSONPath matches no nodes.
+            <t>Syntax errors in the JSONPath SHOULD be detected before matching is attempted since they do not depend on the JSON document.
+               Therefore, an implementation SHOULD take a JSONPath and produce an optional syntax error and then,
+               if and only if an error was not produced, SHOULD take a JSON document and produce a list of matches or an error (but not both).
             </t>
-            <t>The semantics of each matcher are defined for each type of node. Each node holds a JSON value
-               (as defined by <xref target="RFC8259" format="default">RFC&nbsp;8259</xref>), that is, an object, array, number, string, or one
-               of the literals <tt>true</tt>, <tt>false</tt>, or <tt>null</tt>.
+            <t>Alternatively, an implementation MAY take a JSONPath and a JSON document and produce a list of matches or an optional error (but not both).
+            </t>
+            <t>For any implementation, if a syntactically invalid JSONPath is provided, the implementation MUST return an error.
+            </t>
+            <t>If a syntactially invalid JSON document is provided, an implementation MAY return an error or MAY return an empty list of matches.
             </t>
             <section numbered="true" toc="default">
-               <name>Basics</name>
-               <t>JSONPaths start by matching the root node of a JSON document. There is an explicit notation <tt>$</tt> for a matcher which matches just the root node:
+               <name>Result types</name>
+               <t>An implementation SHOULD produce a list of nodes or an error. However, an implementation MAY instead produce a scalar value (a number, a string, or a
+               literal <tt>true</tt>, <tt>false</tt>, or <tt>null</tt>) or an error.
                </t>
-               <sourcecode type="abnf">
-path = root subpath ; 
-root = %x24         ; $ matches the root node of a document
-
-subpath = identity  ; see below for alternatives
-identity = ""       ; matches all input nodes
-               </sourcecode>
-               <t>Valid JSONPaths MUST conform to the grammar of <tt>path</tt>.
+               <t>An implementation which produces a scalar value or an error MUST produce a scalar value only when the list of matching nodes comprises a
```

Comment:
> when the list of matching nodes comprises a single node
This approach can cause many troubles, I think.

Look, there are two cases:
1. Query is able to match zero or one node. It is an _attribute_ of a query, and it can be calculated using just query text, without any JSON data: all matchers in chain should match one or zero nodes (like `.foo') to tag query as "single-matching".
2. Query is able to match arbitrary amount of nodes. It's an attribute of a query, too, and it's enough to have a single "multi-matching" matcher (like `*`) in chain to tag all the query as "multi-matching".

But on some JSON data multi-matching query will match exactly one node, which makes it possible to behave like single-matching query 1; I mean, node list with a single node can be reduced to a single node.

Query attributes can be used for analysis by client software to ensure that given query is garanteed to match either exactly one either none of nodes. Making the type of returned data dependent on input is dangerous and complicated: well, okay, we've got an array here - is this a single node that is an array or is this a list of matched nodes? I see no way to distinguish.

I suggest not to mix "naturally single-matching queries" with "occasionally single-matching queries".

Reply:
Intersting. Let's analyse this a bit. The nub of the problem seems to be:
> Making the type of returned data dependent on input is dangerous and complicated: well, okay, we've got an array here - is this a single node that is an array or is this a list of matched nodes? I see no way to distinguish.

The output type isn't data (JSONPath or input document) dependent _for a given implementation_, but it would be implementation dependent.

Let's look at both implementation options.

Firstly, the preferred (SHOULD) approach of always returning a list of nodes or error. We can overlook the error case as that's unambiguous. If a single node matched, then the output would be a list comprising that node. There is no ambiguity. The "outermost" list always represents the fact the implementation is returning a list of nodes.

Secondly, the non-preferred "scalar" approach of always returning a scalar or error. If a single node matched which contains a scalar value, that value is returned. Again, I don't think there is any ambiguity.

Reply:
But how should "the non-preferred _scalar_ approach" behave if query matches several nodes on concrete data? If it falls back on "preferred approach", the result can be ambiguous. If it doesn't, we should define the behavior better: it can return first node from list, or return an error, or whatever.

Reply:
Ah, I missed the point: you want to allow "single" mode only for scalar values. That indeed removes ambiguity. But in the same time, the non-preferred approach still seems very wierd:

- `$.foo` on `{"foo":"bar"}` returns `"bar"`.
- `$.foo` on `{"foo":["bar","baz"]}` returns `[["bar","baz"]]` - isn't it weird? Or it should produce an error?

What do you think about replacing "non-preferred approach" with optional _reduce_ operation on "preferred approach" results?

Reply:
> Ah, I missed the point: you want to allow "single" mode only for scalar values. That indeed removes ambiguity. But in the same time, the non-preferred approach still seems very wierd:
> 
> * `$.foo` on `{"foo":"bar"}` returns `"bar"`.

Correct.

> * `$.foo` on `{"foo":["bar","baz"]}` returns `[["bar","baz"]]` - isn't it weird? Or it should produce an error?

It should return an error since `[["bar","baz"]]` isn't a "list of matching nodes comprises a single node whose value is [a] scalar value".

> 
> What do you think about replacing "non-preferred approach" with optional _reduce_ operation on "preferred approach" results?

I like the idea of framing the behaviour in terms of _reduce_, but I'm not sure that would provide the "escape route" that I was aiming for. You see I don't personally think the non-preferred approach is a good option, but I suggest it so that implementations which behave this way today could become spec compliant without necessarily completely breaking all their users.

In fact, I put up the current non-preferred approach as a strawman for discussion and I'd be interested in others' opinions (@cburgmer, @gregsdennis, @Cleidiano, and other potential reviewers).

Another, more radical option, would be to drop the non-preferred approach altogether and leave it to implementations to define their behaviour _in terms of the standard_ should they choose to do so.

Another intermediate option would be to add some non-normative text explaining how to return a scalar in terms of the standard behaviour. I quite like that approach as it would at least help "scalar implementations" to adopt a similar approach.

Avoiding a normative definition of scalar implementations would also get us out of the difficulty of covering that use case in a future compliance test suite.

Reply:
> I like the idea of framing the behaviour in terms of reduce, but I'm not sure that would provide the "escape route" that I was aiming for.

Well, it's a big question if existing questions conform to this "escape route", anyway. Just try execute `$.store` [here](https://jsonpath.herokuapp.com/) on Goessner's example - and you'll get a single JSON object. Execute `$.store.book` - and get a single array (exactly what I've meant by "ambiguous"). As for me, I'd vote for the most radical option of dropping the non-preferred approach.

Reduction is a needful operation on practice, though: there are tons of cases when we may want to convert result to exactly one node or error. But I'm not sure if we should bother about such an operation in this document: given a list of nodes there's no problem to implement such a behavior even if some implementation doesn't provide it out of box.

Reply:
I agree. So let's drop it for now and add some _non-normative_ text later if there's any pressure to do so. I'll force push in a few minutes...